### ISSUE TITLE
[HOLD for payment 2024-10-25] [$250] Web - Attachment - Unable to download a txt attachment until the browser is refreshed

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#48679]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.30-7
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Log in with a new Gmail account
2. Navigate to FAB - Start chat
3. Open a chat with any Gmail account
4. Send any txt file
5. Click on the txt file
6. Refresh the browser
7. Click on the txt file

## Expected Result:
I should be able to download the attachment.

## Actual Result:
Unable to download a txt attachment until the browser is refreshed. The download icon and the function itself becomes available again.

## Workaround:
Unknown

## Platforms:

- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos
https://github.com/user-attachments/assets/eba88040-eed4-4685-b08c-3baebe5dc007

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021831987490764045084</li>
        <li>Upwork Job ID: 1831987490764045084</li>
        <li>Last Price Increase: 2024-09-20</li>
<li>Automatic offers: </li>
<ul>
<li>HezekielT | Contributor | 104164554</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @jliexpensify</details>

**Comment by IuliiaHerets** [comment]
@jliexpensify  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by jliexpensify** [comment]
I can reproduce, but I'm only on v9.0.29-12 and I can't actually download the .txt file, even after refreshing multiple times 😡

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Can't download a text file while uploading.

### What is the root cause of that problem?
When we upload the attachment, the source is still a local file (blob://...), so, the sourceID here is empty.
https://github.com/Expensify/App/blob/4af620b439636f714ff72a26320ed57f9d2faac3/src/components/AnchorForAttachmentsOnly/BaseAnchorForAttachmentsOnly.tsx#L35

The sourceID checks for the report action ID from the uploaded attachment source (/chat-attachments/actionID/...). Because the id is empty, the download icon won't be shown and pressing the attachment won't do anything.
https://github.com/Expensify/App/blob/4af620b439636f714ff72a26320ed57f9d2faac3/src/components/AnchorForAttachmentsOnly/BaseAnchorForAttachmentsOnly.tsx#L66
https://github.com/Expensify/App/blob/4af620b439636f714ff72a26320ed57f9d2faac3/src/components/AnchorForAttachmentsOnly/BaseAnchorForAttachmentsOnly.tsx#L47-L50

### What changes do you think we should make in order to solve the problem?

Just want to provide an alternative if we want to allow the user to download the text attachment while still uploading. To do that, instead of finding the action ID from the attachment source, we can find it from the report action itself.
https://github.com/Expensify/App/blob/4af620b439636f714ff72a26320ed57f9d2faac3/src/components/AnchorForAttachmentsOnly/BaseAnchorForAttachmentsOnly.tsx#L43-L44

We also need to convert the `withOnyx` with `useOnyx`.
```
const { anchor, report, reportNameValuePairs, action, checkIfContextMenuActive } = useContext(ShowContextMenuContext);
const sourceID = action?.reportActionID;

const [download] = useOnyx(`${ONYXKEYS.COLLECTION.DOWNLOAD}${sourceID}`);
```
Last, we need to update the source so it only append encryptedAuthToken if it's a remote file.

```
const sourceURLWithAuth = (source.startsWith('blob:') || source.startsWith('file:')) ? source : addEncryptedAuthTokenToURL(source);
```

**Comment by huult** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Unable to download a txt attachment until the browser is refreshed

### What is the root cause of that problem?

The `reportAction` is not updated in IndexedDB as shown in the image below, even after `addAttachment` completes successfully.

<img width="1708" alt="Screenshot 2024-09-08 at 23 19 06" src="https://github.com/user-attachments/assets/faec5592-8cae-493e-aab7-88855e53c1e8">


The [condition check](https://github.com/Expensify/App/blob/main/src/components/AnchorForAttachmentsOnly/BaseAnchorForAttachmentsOnly.tsx#L48) for `sourceID` returns `undefined`, which prevents the attachment from being downloaded.

https://github.com/Expensify/App/blob/9affb1384815fa8155fe23c523127c1e72b1120d/src/components/AnchorForAttachmentsOnly/BaseAnchorForAttachmentsOnly.tsx#L35-L54

The `sourceID` returns `undefined` due to `const sourceID = (source.match(CONST.REGEX.ATTACHMENT_ID) ?? [])[1];`. This happens because the `source` value is `blob:https://dev.new.expensify.com:8082/7f479bf0-ef65-49f7-8f89-9e1cb1bf2d2f`, which is stored in IndexedDB but not updated in `reportAction` even after `addAttachment` completes successfully.

<img width="1709" alt="Screenshot 2024-09-08 at 23 26 55" src="https://github.com/user-attachments/assets/8fad77c0-e4e1-4384-8496-f55fe4eb4053">

### What changes do you think we should make in order to solve the problem?

We should update the `reportAction` data after `addAttachment` completes successfully. Something like this:

```diff
// src/libs/Middleware/SaveResponseInOnyx.ts#L36

+        onyxUpdates.forEach((onyxUpdatesItem) => {
+            const onyxUpdatesKet = onyxUpdatesItem.key;
+            if (onyxUpdatesKet.startsWith(ONYXKEYS.COLLECTION.REPORT_ACTIONS)) {
+                const reportActionId = onyxUpdatesKet.replace(ONYXKEYS.COLLECTION.REPORT_ACTIONS, '');
+
+                if (onyxUpdatesItem.value) {
+                    Onyx.merge(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${reportActionId}`, onyxUpdatesItem.value);
+                }
+            }
+        });
```

<summary>
<details>POC

https://github.com/user-attachments/assets/0e73fa8f-ef1a-4b0e-8019-521503fdc8c2

</details>
</summary>

**Comment by jliexpensify** [comment]
Bumping @getusha to test and look at proposals!

**Comment by getusha** [comment]
I am inclined to @hoangzinh's solution, adding an indicator that the file is still being uploaded sounds good as well.
Probably something like this? (same as when we download after it is uploaded)

![Screenshot 2024-09-10 at 1 07 18 in the afternoon](https://github.com/user-attachments/assets/4f595dd7-d499-498e-b2ce-94c67bb159be)

cc @Expensify/design

**Comment by shawnborton** [comment]
I'm not sure that we need a tooltip on the loading spinner (it's pretty obvious what it means IMO), but I do like the idea of including it until the attachment is ready. That being said, it might make sense to replace the attachment icon with the spinner so that the shape of the attachment doesn't grow/shrink when it loads.

**Comment by huult** [comment]
Hi @getusha Can you check again? The issue with this ticket is not uploading the file, but that the user cannot download the file. The problem is that the user cannot download the file even though it has been successfully uploaded.

https://github.com/user-attachments/assets/92cce693-d7f2-4e2e-8612-58d1b2e55089

**Comment by getusha** [comment]
Hi @huult
I cannot reproduce the issue you mentioned, i tried multiple times and i can download the file after it has been successfully uploaded.

https://github.com/user-attachments/assets/c19ca1b6-fc6f-4ef7-94cc-ef1ff80694b2

**Comment by huult** [comment]
Hi @getusha I had the same problem as you when I first reproduced it. Watch my video below to see how to reproduce the issue (using any email and sending an attachment to any email).

https://github.com/user-attachments/assets/2875b92d-13ef-4ce0-b21c-5256aeacc1a8

**Comment by jliexpensify** [comment]
@getusha bumping this one!

**Comment by getusha** [comment]
@huult any reason why this is happening on newly created chats?

**Comment by jliexpensify** [comment]
Bump @getusha for a review of the above comment, thanks!

**Comment by jliexpensify** [comment]
Bumping @getusha again, and also on Slack

**Comment by llvee** [comment]
@jliexpensify Is this resolved? I am interested in contributing to more projects. I took a look at this, seems it requires testing on multiple platforms, involves many different files without clear direction as to what those are so the bounty does seem low for this if that is true.

**Comment by llvee** [comment]
@MelvinBot Is that response completely automated? Jw for appriopriate followup responses.

**Comment by jliexpensify** [comment]
@llvee all Melvin responses are automated, and it's indicating we don't have your details in our system.

This is still unresolved, we are awaiting a review. if you have a proposal, please follow the proper guidelines in the  Contributing .MD and feel free to post one with the correct format.

**Comment by getusha** [comment]
@huult I am no longer able to reproduce the issue you mentioned

https://github.com/user-attachments/assets/8d3aac98-d77b-4001-8426-7199f523899a

**Comment by getusha** [comment]
I think we just need to show an indicator that the file is still uploading. @hoangzinh could you update your proposal to include that? thanks

**Comment by huult** [comment]
@getusha , Can you watch my video in [this comment](https://github.com/Expensify/App/issues/48679#issuecomment-2340355991)? I am still able to reproduce this issue at the moment!

**Comment by getusha** [comment]
Tested on newly created chat

https://github.com/user-attachments/assets/cce3d617-dbc8-491e-976b-0b90be3e6de3

**Comment by getusha** [comment]
@huult send an attachment to an account that hasn't been created yet?

**Comment by huult** [comment]
@getusha Yes. The same steps as in the [video](https://github.com/Expensify/App/issues/48679#issue-2509735649) for this task.

**Comment by huult** [comment]
@getusha This issue happens with newly registered accounts, and the reason is explained [here](https://github.com/Expensify/App/issues/48679#issuecomment-2352469653). The `reportAction` does not update after calling `AddAttachment`

**Comment by getusha** [comment]
Same @huult 

https://github.com/user-attachments/assets/68fb994b-1710-48ab-99f3-e9ddae21f620

**Comment by huult** [comment]
I was able to reproduce the issue an hour ago, but now I can't reproduce it anymore.

**Comment by llvee** [comment]
I took a look at this again, while I haven't signed up through the channels mentioned. It did become clear after reading through this more, that this likely requires moving away from using local storage to store relevant values related to registering the upload of the txt attachment to leveraging JS code more resilently in order to store, report values using reportAction. 

If this helps out, can I receive a partial bounty if I don't actually create the code before that is completed since others seem to have already been working on this? Since moving away from localstorage & building more resilent JS code is likely a better proposed solution than others.

On that note, those who are already or were already working on this please let me know if you'd like me to contribute a code resolution. I'm interested in helping as mentioned above.

**Comment by jliexpensify** [comment]
@llvee if you are interested, please review our Contributing .MD and submit a proposal using the correct format.

**Comment by llvee** [comment]
@jliexpensify I may not be able to use Upwork presently. Is there any alternative contribution strategy or is the team open to discussion around that? If there isn't, seems important in case of more serious potential development delays due to overreliance on a single platform.

**Comment by getusha** [comment]
@HezekielT's [proposal](https://github.com/Expensify/App/issues/48679#issuecomment-2377484358) looks good to me. let's make sure to disable the hover effect when it's still uploading as @hoangzinh mentioned.
🎀 👀 🎀 C+ Reviewed

**Comment by llvee** [comment]
@getusha according to other comments, the sourceID being null is not due to what is stated in @HezekielT's proposal. I took a look at the video, it only shows uploading the file, the file being successfully uploaded & the loading spinner working which doesn't show a resolution for being able to download the txt attachment without refreshing the browser. Just a heads up about that.

**Comment by aldo-expensify** [comment]
> @jliexpensify I may not be able to use Upwork presently. Is there any alternative contribution strategy or is the team open to discussion around that? If there isn't, seems important in case of more serious potential development delays due to overreliance on a single platform.

The [#expensify-open-source](https://expensify.enterprise.slack.com/archives/C01GTK53T8Q) slack channel may be a better place to discuss this

**Comment by jliexpensify** [comment]
Just a heads up I'm OOO from 3rd to 14th, but I don't think anything is needed from me during this period. Feel free to reassign to another B0 if needed for any payments though!

**Comment by jliexpensify** [comment]
Bump @getusha - any checklist needed?

**Comment by jliexpensify** [comment]
Bumping @getusha for the checklist

Payment Summary

- C: @HezekielT $250 (to be paid via Upworks) - please accept the offer
- C+: @getusha $250 (to be paid via New.Dot)

[Upwork job](https://www.upwork.com/jobs/~021831987490764045084)

**Comment by jliexpensify** [comment]
@HezekielT the job expired, so I had to rehire you [here](https://www.upwork.com/jobs/~021850729658330281708). Please accept.

@getusha - waiting on checklist please.

**Comment by jliexpensify** [comment]
Paid and job closed. 

@getusha [please refer to this summary](https://github.com/Expensify/App/issues/48679#issuecomment-2436724823) for your payment.

**Comment by garrettmknight** [comment]
$250 approved for @getusha
